### PR TITLE
Add logic for active_low pins

### PR DESCRIPTION
**Pull request by ioantotalca**
Add delay for setDirection, adding logic for active_low pins.

### CODE REVIEW

**Comment by AndrewCarterUK** [comment]
Thanks for this!

Comments inline :)

File: src/GPIOInterface.php [hunk 1]
```diff
@@ -19,12 +19,15 @@ public function getInputPin($number);
...
      * Get an output pin.
-     * 
-     * @param int $number The pin number
-     * 
-     * @return OutputPinInterface
+     *
+     * @param int  $number The pin number
+     *
+     * @param bool $activeLow
+     * @param null $defaultHigh
```
Comment (AndrewCarterUK) [inline src/GPIOInterface.php]: `null` isn't really a data type - is this a boolean?

If so, should the default value be false?

File: src/Pin/OutputPin.php [hunk 2]
```diff
@@ -8,15 +8,28 @@
 {
     /**
      * Constructor.
-     * 
+     *
      * @param FileSystemInterface $fileSystem An object that provides file system access
      * @param int                 $number     The number of the pin
+     * @param bool|int            $activeLow
```
Comment (AndrewCarterUK) [inline src/Pin/OutputPin.php]: Think this should just be a bool 'true' or 'false.

File: src/Pin/OutputPin.php [hunk 3]
```diff
@@ -8,15 +8,28 @@
...
-    public function __construct(FileSystemInterface $fileSystem, $number)
+    public function __construct(FileSystemInterface $fileSystem, $number, $activeLow = Pin::ACTIVE_LOW_FALSE, $defaultHigh = null)
     {
         parent::__construct($fileSystem, $number);
 
-        $this->setDirection(self::DIRECTION_OUT);
+        if ($activeLow && !isset($defaultHigh)) {
+            $defaultHigh = true;
+        }
+        if ($defaultHigh) {
```
Comment (AndrewCarterUK) [inline src/Pin/OutputPin.php]: Sorry to be fussy but I'm a blank line between if blocks guy :p

File: src/Pin/Pin.php [hunk 4]
```diff
@@ -69,7 +76,12 @@ public function unexport()
     protected function setDirection($direction)
     {
         $directionFile = $this->getPinFile(self::GPIO_PIN_FILE_DIRECTION);
-        $this->fileSystem->putContents($directionFile, $direction);
+        try {
+            $this->fileSystem->putContents($directionFile, $direction);
+        } catch (\RuntimeException $ex) {
+            usleep(100000);
+            $this->fileSystem->putContents($directionFile, $direction);
+        }
```
Comment (AndrewCarterUK) [inline src/Pin/Pin.php]: Just wondering if there's a neater way of doing this. I like the idea of retrying a fail - can you think of a way of retrying such that we don't retry for any RuntimeException?

On one hand, if it's a permission denied error there's no point retrying at all. On the other hand, if it's a timing issue it makes sense to maybe even try again sooner (more often).

**Comment by AndrewCarterUK** [comment]
Another note, would to be possible to update add in unit tests to obtaining full coverage [see here](https://scrutinizer-ci.com/g/PiPHP/GPIO/inspections/b9fe73e2-be48-4320-bfb0-a2ae2af4c3a3/code-structure/test-coverage).

It's not a catch all - but it helps prevent things from breaking in future changes! :)

Thanks again

**Comment by svpernova09** [comment]
Marking inactive. If this is still valuable please feel free to reopen and address comments. Thanks!

**Pull request closed without merging**